editor/osm_feature_matcher.cpp

```diff
@@ -28,60 +189,98 @@ double ScoreLatLon(XMLFeature const & xmlFt, ms::LatLon const & latLon)
   return 1.0 - (a.Length(b) / kPointDiffEps);
 }
 
-template <typename TFunc>
-void ForEachWaysNode(pugi::xml_document const & osmResponse, pugi::xml_node const & way,
-                     TFunc && func)
+void ForEachRefInWay(pugi::xml_document const & osmResponse, pugi::xml_node const & way,
+                     ForEachRefFn const & fn)
 {
   for (auto const xNodeRef : way.select_nodes("nd/@ref"))
   {
     string const nodeRef = xNodeRef.attribute().value();
     auto const node = osmResponse.select_node(("osm/node[@id='" + nodeRef + "']").data()).node();
     ASSERT(node, ("OSM response have ref", nodeRef, "but have no node with such id.", osmResponse));
     XMLFeature xmlFt(node);
-    func(xmlFt);
+    fn(xmlFt);
   }
 }
 
-template <typename TFunc>
-void ForEachRelationsNode(pugi::xml_document const & osmResponse, pugi::xml_node const & relation,
-                          TFunc && func)
+void ForEachWayInRelation(pugi::xml_document const & osmResponse, pugi::xml_node const & relation,
+                          ForEachWayFn const & fn)
 {
-  for (auto const xNodeRef : relation.select_nodes("member[@type='way']/@ref"))
+  auto const nodesSet = relation.select_nodes("member[@type='way']/@ref");
+  for (auto const & xNodeRef : nodesSet)
   {
     string const wayRef = xNodeRef.attribute().value();
     auto const xpath = "osm/way[@id='" + wayRef + "']";
-    auto const way = osmResponse.select_node(xpath.data()).node();
-    // Some ways can be missed from relation.
-    if (!way)
+    auto const way = osmResponse.select_node(xpath.c_str()).node();
+
+    auto const rolePath = "member[@ref='" + wayRef + "']/@role";
+    pugi::xpath_node role = relation.select_node(rolePath.c_str());
+
+    // Some ways can be missed from relation and
+    // we need to understand role of the way (inner/outer).
+    if (!way || (!role && nodesSet.size() != 1))
       continue;
-    ForEachWaysNode(osmResponse, way, forward<TFunc>(func));
+
+    fn(way, role.attribute().value());
   }
 }
 
-vector<m2::PointD> GetWaysGeometry(pugi::xml_document const & osmResponse,
-                                   pugi::xml_node const & way)
+Polygon GetWaysGeometry(pugi::xml_document const & osmResponse, pugi::xml_node const & way)
 {
-  vector<m2::PointD> result;
-  ForEachWaysNode(osmResponse, way, [&result](XMLFeature const & xmlFt)
-                  {
-                    result.push_back(xmlFt.GetMercatorCenter());
-                  });
+  Polygon result;
+  ForEachRefInWay(osmResponse, way, [&result](XMLFeature const & xmlFt)
```

Comment:
Could you please make a template function for this to avoid duplication?
```
template<typename Geometry> 
GetGeometry()
{
  Geometry result;
  ForEachRefInWay ...
}

GetGeometry<Polygon>(...);
GetGeometry<MultiLinestring>(...);
```

Reply:
готово